Entities/AtomGroup.cpp

```diff
@@ -1,2618 +1,1750 @@
+		bool rayHit = false;
+
+		Vector atomOffset = Vector();
+		Vector atomPos = Vector();
+
+		// First go through all Atoms to find the first intersection and get the intersected MO
+		for (Atom *atom : m_Atoms) {
+			atomOffset = m_OwnerMO->RotateOffset(atom->GetOffset());
+			atom->SetupPos(position + atomOffset);
+			atomPos = atom->GetCurrentPos();
+			hitMaterial = g_SceneMan.GetTerrain()->GetPixel(atomPos.GetFloorIntX(), atomPos.GetFloorIntY());
+			if (hitMaterial != g_MaterialAir && strengthThreshold > 0.0F && g_SceneMan.GetMaterialFromID(hitMaterial)->GetIntegrity() > strengthThreshold) {
+				intersectingAtoms.push_back(atom);
 			}
 		}
-    }
+		if (intersectingAtoms.empty()) {
+			return true;
+		}
 
-    // If no intersected MO is found, we are clear and can quit
-	if (!pIntersectedMO) {
-		return false;
-	}
+		// If all Atoms are intersecting, we're screwed?!
+		if (intersectingAtoms.size() >= m_Atoms.size()) {
+			return false;
+		}
+
+		Vector exitDirection = Vector();
+
+		// Go through all intersecting Atoms and find their average inverse normal
+		for (const Atom *intersectingAtom : intersectingAtoms) {
+			exitDirection += m_OwnerMO->RotateOffset(intersectingAtom->GetNormal());
+		}
+
+		// We don't have a direction to go, so quit
+		// TODO: Maybe use previous position to create an exit direction instead then?
+		if (exitDirection.IsZero()) {
+			return false;
+		}
+
+		// Invert and set appropriate length
+		exitDirection = -exitDirection;
+		exitDirection.SetMagnitude(m_OwnerMO->GetDiameter());
+
+		// See which of the intersecting Atoms has the longest to travel along the exit direction before it clears
+		float longestDistance = 0.0F;
+
+		Vector clearPos = Vector();
+		Vector atomExitVector = Vector();
+		Vector totalExitVector = Vector();
 
-    // Tell both MO's that they have hit an MO, and see if they want to continue
-	if (m_pOwnerMO->OnMOHit(pIntersectedMO)) {
-		return false;
+		for (const Atom *intersectingAtom : intersectingAtoms) {
+			atomPos = intersectingAtom->GetCurrentPos();
+
+			if (strengthThreshold <= 0.0F) {
+				rayHit = g_SceneMan.CastMaterialRay(atomPos, exitDirection, g_MaterialAir, clearPos, 0, false);
+			} else {
+				rayHit = g_SceneMan.CastWeaknessRay(atomPos, exitDirection, strengthThreshold, clearPos, 0, false);
+			}
+
+			if (rayHit) {
+				// Determine the longest clearing distance so far
+				atomExitVector = clearPos - atomPos;
+				if (atomExitVector.GetMagnitude() > longestDistance) {
+					// We found the Atom with the longest to travel along the exit direction to clear, so that's the distance to move the whole object to clear all its Atoms.
+					longestDistance = atomExitVector.GetMagnitude();
+					totalExitVector = atomExitVector;
+				}
+			}
+		}
+
+		// If the exit vector is too large, then avoid the jarring jump and report that we didn't make it out
+		if (totalExitVector.GetMagnitude() > m_OwnerMO->GetRadius()) {
+			//position += totalExitVector / 2;
```

Comment:
Would this commented code have done anything and have you tried it? Either try it and use it or delete it or, if you really don't want to, put a TODO about it.
Also, hey, this is the first comment that I'd say makes sense to keep, hurray!